### PR TITLE
Improve DataSet API documentation with YAML examples

### PR DESCRIPTION
**Pull request by levimjoseph**
## Description
Adding YAML documentation
In response to:  https://github.com/kedro-org/kedro/issues/1762

## Development notes
I changed the following files:
- kedro/extras/datasets/api
- kedro/extras/datasets/dask
- kedro/extras/datasets/pandas
- kedro/extras/datasets/tensorflow
- kedro/extras/datasets/tracking
- kedro/extras/datasets/tracking
- kedro/extras/datasets/yaml

I ran the test suite.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1844"></a>

### CODE REVIEW

**Comment by levimjoseph** [comment]
@noklam this is my draft pr. First PR, so any advice/input is welcome

File: kedro/extras/datasets/tensorflow/tensorflow_model_dataset.py [hunk 1]
```diff
@@ -25,14 +25,28 @@ class TensorFlowModelDataset(AbstractVersionedDataSet[tf.keras.Model, tf.keras.M
...
+
     Example:
     ::
 
         >>> from kedro.extras.datasets.tensorflow import TensorFlowModelDataset
         >>> import tensorflow as tf
         >>> import numpy as np
         >>>
-        >>> data_set = TensorFlowModelDataset("saved_model_path")
+        >>> data_set = TensorFlowModelDataset("saved_model/my_model")
```
Comment (noklam) [inline kedro/extras/datasets/tensorflow/tensorflow_model_dataset.py:49]:
```suggestion
        >>> data_set = TensorFlowModelDataset("data/06_models/tensorflow_model.h5")
```

Looks like the default is a `h5` file, would need to update the `filepath` in the YAML as `h5` as well.

https://www.tensorflow.org/api_docs/python/tf/keras/models/save_model

File: kedro/extras/datasets/yaml/yaml_dataset.py [hunk 2]
```diff
@@ -21,6 +21,24 @@ class YAMLDataSet(AbstractVersionedDataSet[Dict, Dict]):
...
+        >>>   type: yaml.YAMLDataSet
+        >>>   filepath: cars.yaml
+        >>>
+        >>> motorbikes:
+        >>>   type: yaml.YAMLDataSet
+        >>>   filepath: s3://your_bucket/data/02_intermediate/company/motorbikes.csv
+        >>>   credentials: dev_s3
+        >>>
+
+
```
Comment (noklam) [inline kedro/extras/datasets/yaml/yaml_dataset.py:41]: Is this needed?
```suggestion
```

File: kedro/extras/datasets/dask/parquet_dataset.py [hunk 3]
```diff
@@ -15,6 +15,25 @@ class ParquetDataSet(AbstractDataSet[dd.DataFrame, dd.DataFrame]):
...
+
+        >>> cars:
+        >>>   type: dask.ParquetDataSet
+        >>>   filepath: s3://bucket_name/path/to/folder
+        >>>   save_args:
+        >>>     compression: GZIP
+        >>>   credentials:
+        >>>     client_kwargs:
+        >>>         aws_access_key_id: YOUR_KEY
+        >>>         aws_secret_access_key: YOUR SECRET
```
Comment (noklam) [inline kedro/extras/datasets/dask/parquet_dataset.py:33]:
```suggestion
        >>>         aws_secret_access_key: YOUR_SECRET
```

I think the Python API should be update to the same style as well here, just to be more consistent.

**Review by noklam — COMMENTED** [review]
Thanks @levimjoseph for the PR and sorry for the really late review.

Just to check if you have run some of the code to check if the YAML API works?

I notice there are some inconsistencies (some of them are inherit from our docs). In general, the Python API and YAML API are the describing the same dataset, which I like because it demonstrate how to map these arguments clearly. 

In some cases the original example are too simple, so I think it's fine even if the new YAML API are not exactly the same.

There are way too many `s3`, `gcs` paths that are too complicated as an exampe, but we will address this later. For API docs I think it's nice to have at least 1 example that is easy to copy & paste without too many setups.

File: kedro/extras/datasets/api/api_dataset.py [hunk 4]
```diff
@@ -13,6 +13,26 @@ class APIDataSet(AbstractDataSet[None, requests.Response]):
...
+        data_catalog.html#using-the-data-catalog-with-the-yaml-api>`_:
+
+    .. code-block:: yaml
+
+        >>> usda:
+        >>>   type: api.APIDataSet
+        >>>   url: https://quickstats.nass.usda.gov
+        >>>   params:
+        >>>     key: SOME_TOKEN,
+        >>>     format: SON,
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/api/api_dataset.py:28]:
```suggestion
        >>>     format: JSON,
```

File: kedro/extras/datasets/dask/parquet_dataset.py [hunk 5]
```diff
@@ -61,7 +80,6 @@ def __init__(
             filepath: Filepath in POSIX format to a parquet file
                 parquet collection or the directory of a multipart parquet.
             load_args: Additional loading options `dask.dataframe.read_parquet`:
-                https://docs.dask.org/en/latest/generated/dask.dataframe.read_parquet.html
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/dask/parquet_dataset.py:64]: Why did you remove this link?

File: kedro/extras/datasets/tracking/json_dataset.py [hunk 6]
```diff
@@ -14,6 +14,19 @@ class JSONDataSet(JDS):
...
+    Example adding a catalog entry with
+    `YAML API
+    <https://kedro.readthedocs.io/en/stable/data/\
+        data_catalog.html#using-the-data-catalog-with-the-yaml-api>`_:
+
+    .. code-block:: yaml
+
+        >>> cars:
+        >>>   type: tracking.JSONDataSet
+        >>>   filepath: cars.json
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/tracking/json_dataset.py:26]:
```suggestion
        >>>   filepath: data/09_tracking/cars.json
```

This path is consistent with what we recommend in the experiment tracking tutorial: https://kedro.readthedocs.io/en/stable/tutorial/set_up_experiment_tracking.html

File: kedro/extras/datasets/tracking/metrics_dataset.py [hunk 7]
```diff
@@ -16,6 +16,19 @@ class MetricsDataSet(JSONDataSet):
...
+Example adding a catalog entry with
+    `YAML API
+    <https://kedro.readthedocs.io/en/stable/data/\
+        data_catalog.html#using-the-data-catalog-with-the-yaml-api>`_:
+
+    .. code-block:: yaml
+
+        >>> cars:
+        >>>   type: metrics.MetricsDataSet
+        >>>   filepath: cars.json
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/tracking/metrics_dataset.py:28]:
```suggestion
        >>>   filepath: data/09_tracking/cars.json
```

**Review by MerelTheisenQB — COMMENTED** [review]
Thanks for taking this on @levimjoseph 👍 I've left some comments.

**Comment by MerelTheisenQB** [comment]
Hi @levimjoseph , do you need some help finishing this PR?

**Comment by merelcht** [comment]
Hi @levimjoseph do you still want to complete this PR, or should someone from the team take-over? We'd like to get all PRs related to datasets to be merged soon now we're moving our datasets code to a different package (see our [medium blog post](https://medium.com/quantumblack/keeping-up-with-kedro-the-latest-developments-in-our-development-workflow-framework-cbcc415eea9c) for more details)

File: kedro/extras/datasets/tensorflow/tensorflow_model_dataset.py [hunk 8]
```diff
@@ -25,14 +25,28 @@ class TensorFlowModelDataset(AbstractVersionedDataSet[tf.keras.Model, tf.keras.M
     The underlying functionality is supported by, and passes input arguments through to,
     TensorFlow 2.X load_model and save_model methods.
 
+    .. code-block:: yaml
+
+        >>> tensorflow_model:
+        >>>   type: tensorflow.TensorFlowModelDataset
+        >>>   filepath: data/06_models/tensorflow_model
```
Comment (merelcht) [inline kedro/extras/datasets/tensorflow/tensorflow_model_dataset.py:32]:
```suggestion
        >>>   filepath: data/06_models/tensorflow_model.h5
```

File: kedro/extras/datasets/tracking/metrics_dataset.py [hunk 9]
```diff
@@ -20,6 +20,19 @@ class MetricsDataSet(JSONDataSet):
...
+    `YAML API
+    <https://kedro.readthedocs.io/en/stable/data/\
+        data_catalog.html#using-the-data-catalog-with-the-yaml-api>`_:
+
+    .. code-block:: yaml
+
+        >>> cars:
+        >>>   type: metrics.MetricsDataSet
+        >>>   filepath: data/09_tracking/cars.json
+        >>>
```
Comment (merelcht) [inline kedro/extras/datasets/tracking/metrics_dataset.py:33]:
```suggestion

```

**Review by noklam — APPROVED** [review]
Looks good ⭐️ Thank you @levimjoseph and @merelcht

**Review by AhdraMeraliQB — APPROVED** [review]
